homeassistant/components/synology_dsm/common.py

```diff
@@ -251,22 +250,30 @@ async def async_unload(self) -> None:
             # ignore API errors during logout
             pass
 
-    async def async_update(self, now: timedelta | None = None) -> None:
+    async def async_update(self, first_setup: bool = False) -> None:
         """Update function for updating API information."""
         LOGGER.debug("Start data update for '%s'", self._entry.unique_id)
         self._async_setup_api_requests()
         try:
             await self._hass.async_add_executor_job(
                 self.dsm.update, self._with_information
             )
-        except (SynologyDSMLoginFailedException, SynologyDSMRequestException) as err:
-            LOGGER.warning(
-                "Connection error during update, fallback by reloading the entry"
-            )
+        except (
+            SynologyDSMLoginFailedException,
+            SynologyDSMRequestException,
+            SynologyDSMAPIErrorException,
+        ) as err:
             LOGGER.debug(
                 "Connection error during update of '%s' with exception: %s",
                 self._entry.unique_id,
                 err,
             )
+
+            if first_setup:
+                raise err
+
+            LOGGER.warning(
+                "Connection error during update, fallback by reloading the entry"
+            )
             await self._hass.config_entries.async_reload(self._entry.entry_id)
```

Comment:
this is done, to re-setup (those re-authenticate) the API connection after a possible reboot of the NAS

Reply:
Is there a specific reason the library can't try to re-authenticate the api connection automatically once before raising ?

Reply:
to be honest, I'm not 100% sure, but I know, this approach works quite well in the past 🙈 
but yes I fully agree with you, this needs to be improved (_but in a separate PR, which may also need adjustments in the library_)

Reply:
Reloading the integration in the middle of an update is going to be race prone.  We really should try to avoid that; however, it's what is here now, so given that, your fix looks fine for the problem at hand.

Reply:
In the future, checkout #73145 and #72636 for some history on how this can go wrong.

Reply:
I noticed a few issues with the coordinators while reviewing this so I cleaned them up in https://github.com/home-assistant/core/pull/73257

I did not address the reload issue though.